daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueHasher.scala

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) 2019 The DAML Authors. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.digitalasset.daml.lf
+package value
+
+import java.nio.ByteBuffer
+import java.security.MessageDigest
+
+import com.digitalasset.daml.lf.data.{Numeric, Ref, Utf8}
+import com.digitalasset.daml.lf.value.Value._
+
+abstract class ValueHasher {
+
+  private def encode(a: Array[Byte]) =
+    Ref.LedgerString.assertFromString(
+      a.foldLeft(StringBuilder.newBuilder)((acc, b) => acc.append("%02x".format(b))).toString()
+    )
+
+  def hashValue(value: Value[AbsoluteContractId]): Array[Byte]
+
+  final def hashValueString(value: Value[AbsoluteContractId]): Ref.LedgerString =
+    encode(hashValue(value))
+
+  def hashContract(value: Value[AbsoluteContractId], identifier: Ref.Identifier): Array[Byte]
+
+  final def hashContractString(
+      identifier: Ref.Identifier,
+      value: Value[AbsoluteContractId]
+  ): Ref.LedgerString =
+    encode(hashContract(value, identifier))
+
+  def hashContractKey(identifier: Ref.Identifier, value: Value[AbsoluteContractId]): Array[Byte]
+
+  final def hashContractKeyString(
+      identifier: Ref.Identifier,
+      value: Value[AbsoluteContractId]
+  ): Ref.LedgerString =
+    encode(hashContractKey(identifier, value))
+
+}
+
+object ValueHasher extends ValueHasher {
+
+  // tags are used to avoid hash collisions due to equal encoding for different objects
+  private val tagUnit: Byte = 1
+  private val tagTrue: Byte = 2
+  private val tagFalse: Byte = 3
+  private val tagInt64: Byte = 4
+  private val tagNumeric: Byte = 5
+  private val tagDate: Byte = 6
+  private val tagTimeStamp: Byte = 7
+  private val tagText: Byte = 8
+  private val tagParty: Byte = 9
+  private val tagContractId: Byte = 10
+  private val tagNone: Byte = 11
+  private val tagSome: Byte = 12
+  private val tagList: Byte = 13
+  private val tagTextMap: Byte = 14
+  private val tagGenMap: Byte = 15
+  private val tagRecord: Byte = 16
+  private val tagVariant: Byte = 17
+  private val tagEnum: Byte = 18
+  private val tagContract: Byte = 19
+  private val tagContractKey: Byte = 20
+
+  // used for tagging end of variable length data structure
+  private val tagEnd: Byte = -1 // Does not appear in any valid UTF8 string
+
+  private implicit class MessageDigestOp(val digest: MessageDigest) extends AnyVal {
+
+    def mixByte(v: Byte): MessageDigest = {
```

Comment:
Variously

```suggestion
    def mixByte(v: Byte): digest.type = {
```

If we took this _really_ seriously we'd have to [handwrite the implicit conversion](https://bazaar.launchpad.net/~scompall/stylish-types/trunk/view/head:/src/test/scala/com/nocandysw/stylishtypes/check/CheckProperties.scala#L46), but there's probably no need to go that far here, just do what compiles.

Reply:
I have change a bit the code. This does not work anymore.